Add win_iis_webhandler

arestarh opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Adds support for managing HTTP handlers in IIS for a Windows host.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
win_iis_webhandler
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
ansible 2.6.4
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/andriibi/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Sep 15 2016, 22:37:39) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

arestarh committed:
Added: win_iis_webhandler module to manage HTTP handlers in IIS.

arestarh committed:
Documentation fixes.

arestarh committed:
Update win_iis_webhandler.py

arestarh committed:
Update win_iis_webhandler.py

webknjaz commented:
@equelin did you change your github handle?

equelin commented:
I don't know why and how to solve that !

webknjaz commented:
@equelin it's probably listed in some of the modules you contributed to.

webknjaz commented:
Oh, it's here: https://github.com/ansible/ansible/blob/a8d4bf86421d151d8df7132e8e87d04b6662f45a/lib/ansible/modules/windows/win_pester.py#L32

equelin commented:
My bad, didn't know that it has to be my GitHub username

webknjaz commented:
No worries, now you know :)
I've raised a discussion internally to do better validation of those

gundalow commented on lib/ansible/modules/windows/win_iis_webhandler.ps1:
```diff
@@ -0,0 +1,254 @@
+    $result.state = "absent"
+}
+
+Exit-Json -obj $result
```
Missing new line

arestarh committed:
Fixed: missing new line.

jborean93 commented on lib/ansible/modules/windows/win_iis_webhandler.py:
```diff
@@ -0,0 +1,204 @@
+notes:
+- This must be run on a host that has the WebAdministration powershell module installed.
+- Works with Windows Server 2012 and newer.
+- All parameter' description and other techical information was taken from
```
stray single quote

arestarh replied:
Should be fixed now.

jborean93 commented on lib/ansible/modules/windows/win_iis_webhandler.py:
```diff
@@ -0,0 +1,204 @@
+EXAMPLES = r'''
+- name: Create http handler to process php files
+  win_iis_webhandler:
+    name: 'php-fastcgi'
```
No need to quote most of these valuev in the examples unless they are needed. We try to avoid doing this for the examples.

arestarh replied:
Should be fixed now.

jborean93 commented on lib/ansible/modules/windows/win_iis_webhandler.ps1:
```diff
@@ -0,0 +1,254 @@
+}
+catch {
+    $ErrorMessage = "Failed to get info about webhandler '$name' in IIS on '$iispath_for_messages' path (during creating the result object): $($_.Exception.Message)"
+    Fail-Json -obj $result -message $ErrorMessage
```
What happens when state is absent or we are in check mode and "created" a new handler? Doesn't seem like we should be failing here?

arestarh replied:
Should be fixed now.

jborean93 reviewed (dismissed):
Sorry for taking so long to review this, have had a brief look over and just needs 2 things;

* Integration tests
* Change some of the module options to be more snake_case like, e.g. `applicationname` to `application_name`, `allowpathinfo` to `allow_path_info`. Pretty much any multi word ones should be split with `_`.

arestarh committed:
Update win_iis_webhandler.py

arestarh committed:
Update win_iis_webhandler.py

arestarh committed:
Fixes for module itself

Fixes for module itself according to jborean93 suggestions:
- Last try\catch block is not needed.
- Some parameter names were changed to be more snake_case.

arestarh committed:
Fixes for module documentation

Fixes for module documentation according to jborean93 suggestions:
- Some parameter names were changed to be more snake_case.

dagwieers commented:
During the [next Windows WG meeting](https://github.com/ansible/community/issues/420) on 2018-01-29 we will review and decide on the inclusion of this module in Ansible v2.8. Your presence in the meeting could be helpful.

*If there's anything you wanted to improve/fix, please do so before the meeting.*

nitzmahone commented:
After some discussion at the Windows Working Group module review, we're a little nervous to take on more IIS modules into the core distribution right now, especially given the state of the other IIS modules we already have. This would be a good candidate for the new collections feature that's coming in Ansible 2.8, which would allow these modules to be hosted on galaxy.ansible.com and easily installed/upated by users on their Ansible controller.

arestarh commented:
> After some discussion at the Windows Working Group module review, we're a little nervous to take on more IIS modules into the core distribution right now, especially given the state of the other IIS modules we already have. This would be a good candidate for the new collections feature that's coming in Ansible 2.8, which would allow these modules to be hosted on galaxy.ansible.com and easily installed/upated by users on their Ansible controller.

Thanks, guys, for review and feedback.
I thought about it and I agree with you. Besides that, unfortunately, I don't have much time to fully support module.

ShachafGoldstein commented:
@arestarh - can you close the PR if it is no longer required?

arestarh commented:
@nitzmahone @jborean93 
Guys, as Mat already mentioned with recently released ansible version 2.8 new feature called Collections has arrived.

I have several questions and maybe you can give me some suggestions what to do.
1. In general, what should I do with module and with PR in particular ?
2. I've tried to look for documentation regarding **Collections** on https://docs.ansible.com/ansible/latest/, but didn't find such. Could you please give me correct link ?

Thanks.

webknjaz commented:
Hi Andrew, it's not yet well-documented but you may take a look at the example (incomplete) dir structure https://github.com/bcoca/collection and https://github.com/ansible/mazer.
Alternatively, wait for the docs and subsequent improvements.

jborean93 commented:
Collections are now more of a first class citizen with Ansible and we have better documentation around it. You can follow the [user guide](https://docs.ansible.com/ansible/devel/user_guide/collections_using.html) which goes into detail on how to install and use collection inside Ansible itself.

The [dev guide](https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html) goes through the process of how to create your own collection as well as publish it to Ansible Galaxy.
